File: pep249_dbapi/tests/integ/test_put_get_source_compression.py
import pytest

from ..compatibility import NEW_DRIVER_ONLY, OLD_DRIVER_ONLY
from .utils_put_get import (
    as_file_uri,
    create_temporary_stage,
    PUT_ROW_SOURCE_IDX,
    PUT_ROW_TARGET_IDX,
    PUT_ROW_SOURCE_COMPRESSION_IDX,
    PUT_ROW_TARGET_COMPRESSION_IDX,
    PUT_ROW_STATUS_IDX,
)
from ..utils import shared_test_data_dir


@pytest.mark.parametrize("compression_type", ["GZIP", "BZIP2", "BROTLI", "ZSTD"])
def test_put_source_compression_auto_detect_standard_types(cursor, compression_type):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_AUTO_DETECT_STANDARD")
    filename, file_path = compressed_test_file(compression_type)

    # Upload the compressed file to the stage with AUTO_DETECT
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=AUTO_DETECT"
    )

    # Verify that the file was uploaded, compression type was detected correctly and the file was not compressed again
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_TARGET_IDX] == filename
    assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == compression_type
    assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == compression_type
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


def test_put_source_compression_auto_detect_deflate(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_AUTO_DETECT_DEFLATE")
    filename, file_path = compressed_test_file("DEFLATE")

    # Upload the compressed file to the stage with AUTO_DETECT
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=AUTO_DETECT"
    )

    row = cursor.fetchone()

    if OLD_DRIVER_ONLY("BC#2"):
        assert row[PUT_ROW_SOURCE_IDX] == filename
        assert row[PUT_ROW_TARGET_IDX] == filename + ".gz"
        assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "NONE"
        assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "GZIP"
        assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"

    if NEW_DRIVER_ONLY("BC#2"):
        assert row[PUT_ROW_SOURCE_IDX] == filename
        assert row[PUT_ROW_TARGET_IDX] == filename
        assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "DEFLATE"
        assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "DEFLATE"
        assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


def test_put_source_compression_auto_detect_none_no_auto_compress(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_AUTO_DETECT_NONE_NO_AUTO_COMPRESS")
    filename, file_path = compressed_test_file("NONE")

    # Upload the file to the stage with AUTO_DETECT and AUTO_COMPRESS=FALSE
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=AUTO_DETECT AUTO_COMPRESS=FALSE"
    )

    # Verify that the file was uploaded, compression type was detected as "NONE" correctly and the file was not compressed
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_TARGET_IDX] == filename
    assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "NONE"
    assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "NONE"
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


def test_put_source_compression_auto_detect_none_with_auto_compress(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_AUTO_DETECT_NONE_WITH_AUTO_COMPRESS")
    filename, file_path = compressed_test_file("NONE")

    # Upload the file to the stage with AUTO_DETECT and AUTO_COMPRESS=TRUE
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=AUTO_DETECT AUTO_COMPRESS=TRUE"
    )

    # Verify that the file was uploaded, compression type was detected as "NONE" correctly and the file was compressed to gzip
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_TARGET_IDX] == f"{filename}.gz"
    assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "NONE"
    assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "GZIP"
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


@pytest.mark.parametrize("compression_type", ["GZIP", "BZIP2", "ZSTD", "DEFLATE", "RAW_DEFLATE"])
def test_put_source_compression_explicit_standard_types(cursor, compression_type):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_EXPLICIT_COMPRESSION")
    filename, file_path = compressed_test_file(compression_type)

    # Upload the compressed file to the stage with the specified compression type
    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION={compression_type}"
    )

    # Verify that the file was uploaded, compression type was detected correctly and the file was not compressed again
    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_TARGET_IDX] == filename
    assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == compression_type
    assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == compression_type
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


def test_put_source_compression_explicit_brotli(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_EXPLICIT_BROTLI")
    filename, file_path = compressed_test_file("BROTLI")

    if OLD_DRIVER_ONLY("BC#3"):
        with pytest.raises(Exception):
            cursor.execute(
                f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=BROTLI"
            )

    if NEW_DRIVER_ONLY("BC#3"):
        cursor.execute(
            f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=BROTLI"
        )
        row = cursor.fetchone()
        assert row[PUT_ROW_SOURCE_IDX] == filename
        assert row[PUT_ROW_TARGET_IDX] == filename
        assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "BROTLI"
        assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "BROTLI"
        assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


def test_put_source_compression_explicit_none_no_auto_compress(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_EXPLICIT_NONE_NO_AUTO_COMPRESS")
    filename, file_path = compressed_test_file("NONE")

    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=NONE AUTO_COMPRESS=FALSE"
    )

    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_TARGET_IDX] == filename
    assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "NONE"
    assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "NONE"
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


def test_put_source_compression_explicit_with_auto_compress(cursor):
    stage_name = create_temporary_stage(cursor, "PYTEST_STAGE_EXPLICIT_WITH_AUTO_COMPRESS")
    filename, file_path = compressed_test_file("NONE")

    cursor.execute(
        f"PUT 'file://{as_file_uri(file_path)}' @{stage_name} SOURCE_COMPRESSION=NONE AUTO_COMPRESS=TRUE"
    )

    row = cursor.fetchone()
    assert row[PUT_ROW_SOURCE_IDX] == filename
    assert row[PUT_ROW_TARGET_IDX] == f"{filename}.gz"
    assert row[PUT_ROW_SOURCE_COMPRESSION_IDX] == "NONE"
    assert row[PUT_ROW_TARGET_COMPRESSION_IDX] == "GZIP"
    assert row[PUT_ROW_STATUS_IDX] == "UPLOADED"


def compression_tests_dir():
    return shared_test_data_dir() / "compression"


def compressed_test_file(compression_type: str):
    ct = compression_type.upper()
    base = compression_tests_dir()
    if ct == "GZIP":
        fn = "test_data.csv.gz"
    elif ct == "BZIP2":
        fn = "test_data.csv.bz2"
    elif ct == "BROTLI":
        fn = "test_data.csv.br"
    elif ct == "ZSTD":
        fn = "test_data.csv.zst"
    elif ct == "DEFLATE":
        fn = "test_data.csv.deflate"
    elif ct == "RAW_DEFLATE":
        fn = "test_data.csv.raw_deflate"
    elif ct == "LZMA":
        fn = "test_data.csv.xz"
    elif ct == "NONE":
        fn = "test_data.csv"
    else:
        raise ValueError(f"Unsupported compression type: {compression_type}")
    return fn, base / fn

File: odbc_tests/tests/put_get_tests/test_put_get_source_compression.cpp
#include <filesystem>
#include <string>

#include <catch2/catch_test_macros.hpp>

#include "Connection.hpp"
#include "get_data.hpp"
#include "put_get_utils.hpp"

namespace fs = std::filesystem;
using namespace pg_utils;

static fs::path compression_tests_dir() { return shared_test_data_dir() / "compression"; }

static std::pair<std::string, fs::path> test_file(const std::string& compression_type) {
  if (compression_type == "GZIP") {
    return {"test_data.csv.gz", compression_tests_dir() / "test_data.csv.gz"};
  } else if (compression_type == "BZIP2") {
    return {"test_data.csv.bz2", compression_tests_dir() / "test_data.csv.bz2"};
  } else if (compression_type == "BROTLI") {
    return {"test_data.csv.br", compression_tests_dir() / "test_data.csv.br"};
  } else if (compression_type == "ZSTD") {
    return {"test_data.csv.zst", compression_tests_dir() / "test_data.csv.zst"};
  } else if (compression_type == "DEFLATE") {
    return {"test_data.csv.deflate", compression_tests_dir() / "test_data.csv.deflate"};
  } else if (compression_type == "RAW_DEFLATE") {
    return {"test_data.csv.raw_deflate", compression_tests_dir() / "test_data.csv.raw_deflate"};
  } else if (compression_type == "LZMA") {
    return {"test_data.csv.xz", compression_tests_dir() / "test_data.csv.xz"};
  } else if (compression_type == "NONE") {
    return {"test_data.csv", compression_tests_dir() / "test_data.csv"};
  }
  FAIL("Unsupported compression type: " << compression_type);
  return {"", ""};
}

TEST_CASE("PUT SOURCE_COMPRESSION=AUTO_DETECT standard types", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_AUTO_DETECT_STD");

  for (const std::string comp : {"GZIP", "BZIP2", "BROTLI", "ZSTD"}) {
    auto [filename, file] = test_file(comp);
    auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                   " SOURCE_COMPRESSION=AUTO_DETECT");

    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename);
    compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), comp);
    compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), comp);
    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
  }
}

TEST_CASE("PUT SOURCE_COMPRESSION=AUTO_DETECT with DEFLATE", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_AUTO_DETECT_DEFLATE");
  auto [filename, file] = test_file("DEFLATE");

  auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                 " SOURCE_COMPRESSION=AUTO_DETECT");

  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename);
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), "DEFLATE");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), "DEFLATE");
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
}

TEST_CASE("PUT SOURCE_COMPRESSION=AUTO_DETECT NONE with AUTO_COMPRESS=FALSE", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_AUTO_DETECT_NONE_NO_AC");
  auto [filename, file] = test_file("NONE");

  auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                 " SOURCE_COMPRESSION=AUTO_DETECT AUTO_COMPRESS=FALSE");

  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename);
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), "NONE");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), "NONE");
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
}

TEST_CASE("PUT SOURCE_COMPRESSION=AUTO_DETECT NONE with AUTO_COMPRESS=TRUE", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_AUTO_DETECT_NONE_AC");
  auto [filename, file] = test_file("NONE");

  auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                 " SOURCE_COMPRESSION=AUTO_DETECT AUTO_COMPRESS=TRUE");

  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename + ".gz");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), "NONE");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), "GZIP");
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
}

TEST_CASE("PUT SOURCE_COMPRESSION= explicit standard types", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_EXPLICIT_STD");

  for (const std::string comp : {"GZIP", "BZIP2", "ZSTD", "DEFLATE", "RAW_DEFLATE"}) {
    auto [filename, file] = test_file(comp);
    auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                   " SOURCE_COMPRESSION=" + comp);

    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename);
    compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), comp);
    compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), comp);
    CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
  }
}

TEST_CASE("PUT SOURCE_COMPRESSION=BROTLI explicit", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_EXPLICIT_BROTLI");
  auto [filename, file] = test_file("BROTLI");

  auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                 " SOURCE_COMPRESSION=BROTLI");

  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename);
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), "BROTLI");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), "BROTLI");
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
}

TEST_CASE("PUT SOURCE_COMPRESSION=NONE with AUTO_COMPRESS=FALSE explicit", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_EXPLICIT_NONE_NO_AC");
  auto [filename, file] = test_file("NONE");

  auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                 " SOURCE_COMPRESSION=NONE AUTO_COMPRESS=FALSE");

  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename);
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), "NONE");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), "NONE");
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
}

TEST_CASE("PUT SOURCE_COMPRESSION=NONE with AUTO_COMPRESS=TRUE explicit", "[put_get][odbc]") {
  Connection conn;
  const std::string stage = create_stage(conn, "ODBCTST_SC_EXPLICIT_NONE_AC");
  auto [filename, file] = test_file("NONE");

  auto stmt = conn.execute_fetch("PUT 'file://" + as_file_uri(file) + "' @" + stage +
                                 " SOURCE_COMPRESSION=NONE AUTO_COMPRESS=TRUE");

  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_IDX) == filename);
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_IDX) == filename + ".gz");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_SOURCE_COMPRESSION_IDX), "NONE");
  compare_compression_type(get_data<SQL_C_CHAR>(stmt, PUT_ROW_TARGET_COMPRESSION_IDX), "GZIP");
  CHECK(get_data<SQL_C_CHAR>(stmt, PUT_ROW_STATUS_IDX) == std::string("UPLOADED"));
}
